marlowe-playground-client/src/SimulationPage/State.purs

```diff
@@ -0,0 +1,303 @@
+import Data.Lens.Extra (peruse)
+import Data.Lens.NonEmptyList (_Head)
+import Data.List.NonEmpty as NEL
+import Data.List.Types (NonEmptyList)
+import Data.Map as Map
+import Data.Maybe (Maybe(..), fromMaybe)
+import Data.NonEmptyList.Extra (tailIfNotEmpty)
+import Data.RawJson (RawJson(..))
+import Data.Traversable (for_)
+import Data.Tuple (Tuple(..))
+import Effect.Aff.Class (class MonadAff)
+import Effect.Class (class MonadEffect, liftEffect)
+import Effect.Console (log)
+import Foreign.Generic (ForeignError, decode)
+import Foreign.JSON (parseJSON)
+import Halogen (HalogenM, get, modify_, query)
+import Halogen.Monaco (Query(..)) as Monaco
+import LocalStorage as LocalStorage
+import MainFrame.Types (ChildSlots, _simulatorEditorSlot)
+import Marlowe (SPParams_)
+import Marlowe as Server
+import Marlowe.Monaco as MM
+import Marlowe.Semantics (ChoiceId(..), Input(..), Party(..), inBounds)
+import Network.RemoteData (RemoteData(..))
+import Network.RemoteData as RemoteData
+import Prelude (Unit, Void, bind, discard, flip, identity, mempty, pure, show, unit, zero, ($), (-), (<), (<$>), (<<<), (<>), (=<<), (==), (>=))
+import Servant.PureScript.Ajax (AjaxError, errorToString)
+import Servant.PureScript.Settings (SPSettings_)
+import SimulationPage.Types (Action(..), ActionInput(..), ActionInputId(..), ExecutionState(..), Parties(..), State, _SimulationNotStarted, _SimulationRunning, _bottomPanelView, _currentContract, _currentMarloweState, _executionState, _helpContext, _initialSlot, _marloweState, _moveToAction, _oldContract, _pendingInputs, _possibleActions, _showBottomPanel, _showRightPanel, emptyExecutionStateWithSlot, emptyMarloweState, mapPartiesActionInput)
+import Simulator (applyInput, inFuture, moveToSignificantSlot, moveToSlot, nextSignificantSlot, updateContractInState, updateMarloweState)
+import StaticData (simulatorBufferLocalStorageKey)
+import Text.Pretty (genericPretty)
+import Types (WebData)
+import Web.DOM.Document as D
+import Web.DOM.Element (setScrollTop)
+import Web.DOM.Element as E
+import Web.DOM.HTMLCollection as WC
+import Web.HTML as Web
+import Web.HTML.HTMLDocument (toDocument)
+import Web.HTML.Window as W
+
+handleAction ::
+  forall m.
+  MonadAff m =>
+  SPSettings_ SPParams_ -> Action -> HalogenM State Action ChildSlots Void m Unit
+handleAction settings Init = do
+  editorSetTheme
+  mContents <- liftEffect $ LocalStorage.getItem simulatorBufferLocalStorageKey
+  handleAction settings $ LoadContract $ fromMaybe "" mContents
+
+handleAction settings (SetInitialSlot initialSlot) = do
+  assign (_currentMarloweState <<< _executionState <<< _SimulationNotStarted <<< _initialSlot) initialSlot
+  setOraclePrice settings
+
+handleAction settings StartSimulation = do
+  maybeInitialSlot <- peruse (_currentMarloweState <<< _executionState <<< _SimulationNotStarted <<< _initialSlot)
+  for_ maybeInitialSlot \initialSlot -> do
+    -- FIXME: there is currently bug with the UNDO button on the first step
```

Comment:
was this an existing bug?

Reply:
yes and no 😅... It didn't manifest visually as it does now, so the user would not perceive it as a bug. Frankly, I'm not sure what change made this visible.

There were a couple of problems. The `hasHistory` function inside the Simulator always returned true, that is why I changed from
```purescript
hasHistory :: forall s. { marloweState :: NonEmptyList MarloweState | s } -> Boolean
hasHistory state = has (_marloweState <<< _Tail) state
```
to
```purescript
hasHistory :: forall s. { marloweState :: NonEmptyList MarloweState | s } -> Boolean
hasHistory state = case state ^. (_marloweState <<< _Tail) of
  Nil -> false
  Cons _ _ -> true
```

That was not visible because there weren't any CSS indication of the button being disabled. Now the button is disabled when `hasHistory` returns false instead of firing Nothing as an action

<img width="504" alt="Screen Shot 2020-12-30 at 13 13 28" src="https://user-images.githubusercontent.com/2634059/103366280-f163bb80-4aa0-11eb-8009-cfb773cc61b2.png">

But the other problem that still persists, and it may make sense to fix in a different PR is that when you start the simulation you have two entries in the NonEmptyList. The first one is the one that causes "the visual bug" that is shown in the screenshot, and the second one is what I think should be the first step.

Reply:
I solved the problem in [this commit](https://github.com/input-output-hk/plutus/pull/2560/commits/4e386afcff50231ef14dbc1abacaad17a61e22ef) but I think we should revisit the StartSimulation, ResetSimulation actions. cc @palas